### PR TITLE
Add `Arith256Memory` machine

### PR DESCRIPTION
**Pull request by georgwiese**
Depends on #2117 

This PR adds a new `Arith256Memory` machine. It is largely a copy of `std::machines::large_field::arith::Arith`, but receives its inputs / writes its outputs via a memory pointer.

The first commit copies the existing machine and test. So, I would recommend reviewing the [diff to the first commit](https://github.com/powdr-labs/powdr/compare/7f2c159f1edfc3c1063ee961f706b2d9a66e0e64...arith-memory).

### CODE REVIEW

File: test_data/std/arith_large_test.asm [hunk 1]
```diff
@@ -267,7 +267,8 @@ machine Main with degree: 65536 {
         assert_eq t_0_0, t_0_1, t_0_2, t_0_3, t_0_4, t_0_5, t_0_6, t_0_7, 0x70afe85a, 0xc5b0f470, 0x9620095b, 0x687cf441, 0x4d734633, 0x15c38f00, 0x48e7561b, 0xd01115d5;
         assert_eq t_1_0, t_1_1, t_1_2, t_1_3, t_1_4, t_1_5, t_1_6, t_1_7, 0xf4062327, 0x6b051b13, 0xd9a86d52, 0x79238c5d, 0xe17bd815, 0xa8b64537, 0xc815e0d7, 0xa9f34ffd;
 
-        // Auto-generated rest of the test cases:
+        // Auto-generated rest of the test cases,
+        // see: https://gist.github.com/georgwiese/a66425ae95c839b548edffffe6a692f9
```
Comment (georgwiese) [inline test_data/std/arith_large_test.asm:271]: Thankfully, I found an old Python notebook that used to generate these tests. I uploaded them to Gist and linked it here for future reference. The script also contains code to generate the memory tests.

File: pipeline/tests/powdr_std.rs [hunk 2]
```diff
@@ -137,6 +137,25 @@ fn arith_large_test() {
...
+    // is too slow for the PR tests. This will only create a single
+    // eStark proof instead of 3.
+    #[cfg(feature = "estark-starky")]
+    pipeline
+        .with_backend(powdr_backend::BackendType::EStarkStarky, None)
+        .compute_proof()
+        .unwrap();
+
+    test_halo2(make_simple_prepared_pipeline(f));
+}
```
Comment (leonardoalt) [inline pipeline/tests/powdr_std.rs:157]: Why all but P3?

File: std/machines/large_field/arith256_memory.asm [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+    operation_id: operation_id,
+    // Allow this machine to be connected via a permutation
+    call_selectors: sel,
+{
+    require_field_bits(48, || "Arith256Memory requires a field that fits any 48-Bit value.");
+
+    Byte2 byte2;
+
+    // One-hot encode the operation
+    pol commit is_affine, is_mod, is_ec_add, is_ec_double;
```
Comment (leonardoalt) [inline std/machines/large_field/arith256_memory.asm:31]: col witness?

**Comment by leonardoalt** [comment]
still needs updates

File: std/machines/large_field/arith256_memory.asm (hunk 3 above)
Comment (georgwiese) [reply std/machines/large_field/arith256_memory.asm]: done

File: pipeline/tests/powdr_std.rs (hunk 2 above)
Comment (georgwiese) [reply pipeline/tests/powdr_std.rs:157]: Sorry, reverted again. It fails because the constraint degree is too large.

Comment (leonardoalt) [reply pipeline/tests/powdr_std.rs:157]: ah true. Does this actually run with Halo2 in CI?

**Review by leonardoalt — APPROVED** [review]

File: pipeline/tests/powdr_std.rs (hunk 2 above)
Comment (georgwiese) [reply pipeline/tests/powdr_std.rs:157]: Yes, think so!

**Pull request merged**